lib/Sema/CSDiagnostics.cpp

```diff
@@ -101,6 +101,52 @@ Expr *FailureDiagnostic::getArgumentExprFor(Expr *anchor) const {
   return nullptr;
 }
 
+ConstraintLocator *
+FailureDiagnostic::getConstraintLocatorFor(Expr *expr) const {
+  auto locator = CS.getConstraintLocator(expr);
+
+  if (isa<DeclRefExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr);
+  } else if (isa<UnresolvedDeclRefExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr);
+  } else if (isa<SubscriptExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr, ConstraintLocator::SubscriptMember);
+  } else if (isa<UnresolvedDotExpr>(expr)) {
+    auto UDE = dyn_cast<UnresolvedDotExpr>(expr);
+    if (isa<ConstructorRefCallExpr>(UDE->getBase())) {
+      locator =
+          CS.getConstraintLocator(expr, ConstraintLocator::ConstructorMember);
+    } else {
+      locator = CS.getConstraintLocator(expr, ConstraintLocator::Member);
+    }
+  } else if (isa<MemberRefExpr>(expr)) {
+    auto MRE = dyn_cast<MemberRefExpr>(expr);
+    if (isa<ConstructorRefCallExpr>(MRE->getBase())) {
+      locator =
+          CS.getConstraintLocator(expr, ConstraintLocator::ConstructorMember);
+    } else {
+      locator = CS.getConstraintLocator(expr, ConstraintLocator::Member);
+    }
+  } else if (isa<CallExpr>(expr)) {
+    auto CE = cast<CallExpr>(expr);
+    auto Fn = CE->getFn();
+    locator = CS.getConstraintLocator(Fn, ConstraintLocator::FunctionResult);
```

Comment:
What you have to do in this case is to recurse into this function and give it `CE->getFn()` expression because actual choice comes from the "function expression" attached to application e.g.

```swift
struct S {
  func foo(_: Int) {}
  func foo(_: String) {}
}

func bar(_ s: S) {
  s.foo(42)
}
```

AST looks like this:

```
(call_expr type='()' location=<REPL Input>:2:5 range=[<REPL Input>:2:3 - line:2:11] arg_labels=_:
  (unresolved_dot_expr type='$T1' location=<REPL Input>:2:5 range=[<REPL Input>:2:3 - line:2:5] field 'foo' function_ref=single
    (declref_expr type='S' location=<REPL Input>:2:3 range=[<REPL Input>:2:3 - line:2:3] decl=REPL_1.(file).bar(_:).s@<REPL Input>:1:12 function_ref=unapplied))
  (paren_expr type='($T2)' location=<REPL Input>:2:9 range=[<REPL Input>:2:8 - line:2:11]
    (integer_literal_expr type='$T2' location=<REPL Input>:2:9 range=[<REPL Input>:2:9 - line:2:9] value=42 builtin_initializer=**NULL** initializer=**NULL**)))
```

and disjunction has following locator:

```
disjunction [[locator@0x7fe797801ea0 [UnresolvedDot@<REPL Input>:2:5 -> member]]]:$T1 bound to decl REPL_0.(file).S.foo@<REPL Input>:1:17 : (S) -> (Int) -> () at <REPL Input>:1:17 [[locator@0x7fe797801ea0 [UnresolvedDot@<REPL Input>:2:5 -> member]]]; or $T1 bound to decl REPL_0.(file).S.foo@<REPL Input>:2:8 : (S) -> (String) -> () at <REPL Input>:2:8 [[locator@0x7fe797801ea0 [UnresolvedDot@<REPL Input>:2:5 -> member]]];
```

Reply:
Note that locator is `[UnresolvedDot@<REPL Input>:2:5 -> member]`